TWTValidation/JSON Validator/Schema Validators/TWTJSONSchemaArrayValidator.m

```diff
@@ -0,0 +1,155 @@
+                        minimumItemCount:(NSNumber *)minimumItemCount
+                     requiresUniqueItems:(BOOL)requiresUniqueItems
+                           itemValidator:(TWTValidator *)itemValidator
+                   indexedItemValidators:(NSArray *)indexedItemValidators
+                additionalItemsValidator:(TWTValidator *)additionalItemsValidator
+{
+    self = [super init];
+    if (self) {
+        _maximumItemCount = maximumItemCount;
+        _minimumItemCount = minimumItemCount;
+        _requiresUniqueItems = requiresUniqueItems;
+        _itemValidator = itemValidator;
+        _indexedItemValidators = [indexedItemValidators copy];
+        _additionalItemsValidator = additionalItemsValidator;
+    }
+    return self;
+}
+
+
+- (NSUInteger)hash
+{
+    return [super hash] ^ self.maximumItemCount.hash ^ self.minimumItemCount.hash ^ self.requiresUniqueItems ^ self.itemValidator.hash ^
+    self.indexedItemValidators.hash ^ self.additionalItemsValidator.hash;
+}
+
+
+- (BOOL)isEqual:(id)object
+{
+    if (![super isEqual:object]) {
+        return NO;
+    } else if (self == object) {
+        return YES;
+    }
+
+    typeof(self) other = object;
+    return [other.maximumItemCount isEqual:self.maximumItemCount] && [other.minimumItemCount isEqual:self.minimumItemCount] &&
+    other.requiresUniqueItems == self.requiresUniqueItems && [other.itemValidator isEqual:self.itemValidator] &&
+    [other.indexedItemValidators isEqual:self.indexedItemValidators] && [other.additionalItemsValidator isEqual:self.additionalItemsValidator];
+
+}
+
+
+- (BOOL)validateValue:(id)value error:(out NSError *__autoreleasing *)outError
+{
+    if (![super validateValue:value error:outError]) {
+        return NO;
+    } else if (![value isKindOfClass:[NSArray class]]) {
+        if (outError) {
+            //            *outError = [NSError twt_validationErrorWithCode:TWTValidationErrorCodeValueHasIncorrectClass
+            //                                            failingValidator:self
+            //                                                       value:value
+            //                                        localizedDescription:TWTLocalizedString(@"TWTJSONSchemaArrayValidator.notArrayError")];
+        }
+        return NO;
+    }
+
+    BOOL countValidated = YES;
+    BOOL uniqueItemsValidated = YES;
+    BOOL itemsValidated = YES;
+    BOOL additionalItemsValidated = YES;
+
+    NSError *countError = nil;
+    NSError *uniqueItemsError = nil;
+    NSMutableArray *itemsErrors = outError ? [[NSMutableArray alloc] init] : nil;
+    NSMutableArray *additionalItemsErrors = outError ? [[NSMutableArray alloc] init] : nil;
+
+    if (self.maximumItemCount || self.minimumItemCount) {
+        TWTValidator *countValidator = [[TWTNumberValidator alloc] initWithMinimum:self.minimumItemCount maximum:self.maximumItemCount];
+        countValidated = [countValidator validateValue:@([value count]) error:&countError];
```

Comment:
Sure, but people can invoke this method many times. We’re creating an object every time when we could just create it once in `‑init` and use it forever. Just do this `if (self.minimumCount || self.maximumCount)` thing in `init`, save it in a property, and check the property before doing the validation here.